src/api.ts

```diff
@@ -0,0 +1,73 @@
+export interface Config {
+    url: string | URL;
+    authToken?: string;
+    transactions?: boolean;
+}
+
+export interface Client {
+    execute(stmt: InStatement): Promise<ResultSet>;
+    batch(stmts: Array<InStatement>): Promise<Array<ResultSet>>;
+    transaction(): Promise<Transaction>;
+    close(): void;
+    closed: boolean;
+}
+
+export interface Transaction {
+    execute(stmt: InStatement): Promise<ResultSet>;
+    rollback(): Promise<void>;
+    commit(): Promise<void>;
+    close(): void;
+    closed: boolean;
+}
+
+export interface ResultSet {
+    columns: Array<string>;
+    rows: Array<Row>;
+    rowsAffected: number;
+}
+
+export interface Row {
+    length: number;
+    [index: number]: Value;
+    [name: string]: Value;
+}
+
+export type Value =
+    | null
+    | string
+    | number
+    | ArrayBuffer
+
+export type InValue =
+    | Value
+    | bigint
+    | boolean
+    | Uint8Array
+    | Date
+
+export type InStatement = { sql: string, args: InArgs } | string;
+export type InArgs = Array<InValue> | Record<string, InValue>;
+
+export class LibsqlError extends Error {
+    code: ErrorCode | undefined;
```

Comment:
What does it mean if this is undefined?  Could we arrange for this to always have a value instead?

Reply:
This happens when we don't have any meaningful error code to return; in particular, when we get an error from Hrana or from the HTTP API, because the protocol does not support error codes yet. Do you think that a catch-all error code such as `UNKNOWN` would be better than `undefined`?

Reply:
Yeah, in general I prefer the idea that errors are never undefined/null. It's also easier to write documentation and sample code for this situation.